Add `toHexString` to `UintUtils` and utilize for `AddressUtils#toString`

ItsNickBarry opened the pull request:
1. Implement missing tests in `test/utils/UintUtils.ts`
2. Create similar test file for `AddressUtils`, implement tests for the `toString` function
3. When tests are passing, replace `AddressUtils#toString` implementation with a call to `UintUtils#toHexString` as described [here](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2197#issuecomment-981513782).

ItsNickBarry commented:
Should also be possible to replace the `while` loop in `UintUtils#toHexString` with this `for` loop:
```
for (uint256 temp = value; temp != 0; temp >>=8) {
  length++;
}
```

Check gas difference after implementing tests.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -31,11 +31,74 @@ describe('UintUtils', function () {
     describe('#toHexString', function () {
       describe('(uint256)', function () {
-        it('todo');
+        it('Should return 0 if input is 0', async () => {
```
```suggestion
        it('returns 0 if input is 0', async () => {
```
The word "should" is often used in test descriptions, but it only carries meaning in the context of test/spec religious dogma.  Better to use the action verb directly.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -31,11 +31,74 @@ describe('UintUtils', function () {
+          ).to.equal('0x00');
+        });
+
+        it('Should return correct hexString represenation of a number', async () => {
```
```suggestion
        it('returns correct hexadecimal string representation of a number', async () => {
```
Spelling, and the word "should" (see above).

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -31,11 +31,74 @@ describe('UintUtils', function () {
 
       describe('(uint256,uint256)', function () {
-        it('todo');
+        it('Fails if length input is 0 and value is nonzero', async () => {
```
```suggestion
        it('fails if length input is 0 and value is nonzero', async () => {
```
Use lower-case test descriptors (for consistency).

ItsNickBarry replied:
Also, I prefer to put revert tests in a `describe('reverts if', function () { ... })` block.

ItsNickBarry commented on contracts/utils/UintUtilsMock.sol:
```diff
@@ -10,4 +10,16 @@ contract UintUtilsMock {
+        pure
+        returns (string memory)
+    {
+        return UintUtils.toHexString(value, length);
```
```suggestion
        return value.toHexString(length);
```
For consistency.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -31,11 +31,74 @@ describe('UintUtils', function () {
+          ).to.be.revertedWith('Strings: hex length insufficient');
+        });
+
+        it('Should return correct hexString representation for matching value and length pairs', async () => {
```
```suggestion
        it('returns hexadecimal string representation for matching value and length pairs', async () => {
```

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -28,5 +28,78 @@ describe('UintUtils', function () {
+    describe('#toHexString', function () {
+      describe('(uint256)', function () {
+        it('returns 0 if input is 0', async () => {
+          const zeroNumber = ethers.BigNumber.from('0');
```
Use `ethers.constants.Zero` for consistency.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -28,5 +28,78 @@ describe('UintUtils', function () {
+      });
+
+      describe('(uint256,uint256)', function () {
+        it('fails if length input is 0 and value is nonzero', async () => {
+          await expect(
+            instance.callStatic['toHexString(uint256,uint256)'](
+              ethers.BigNumber.from('100'),
+              ethers.BigNumber.from('0'),
+            ),
+          ).to.be.revertedWith('Strings: hex length insufficient');
+        });
```
Put revert tests in a `describe('reverts if', ...)` block at the end of the parent describe block.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -99,6 +90,17 @@ describe('UintUtils', function () {
           }
         });
+
+        describe('reverts', () => {
```
```suggestion
        describe('reverts if', () => {
```

NouDaimon replied:
Done.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -99,6 +90,17 @@ describe('UintUtils', function () {
         });
+
+        describe('reverts', () => {
+          it('fails if length input is 0 and value is nonzero', async () => {
```
```suggestion
          it('length input is 0 and value is nonzero', async () => {
```

NouDaimon replied:
Done.

ItsNickBarry commented:
@NouDaimon Address comments only after merging #64 to avoid merge conflicts.

ItsNickBarry commented:
@NouDaimon Address comments only after merging #64 to avoid merge conflicts.

ItsNickBarry commented on contracts/utils/UintUtils.sol:
```diff
@@ -26,4 +32,39 @@ library UintUtils {
+            value >>= 4;
+        }
+
+        require(value == 0, 'Strings: hex length insufficient');
```
Prefix error string with "UintUtils", not "Strings".

NouDaimon commented on test/utils/AddressUtils.ts:
```diff
@@ -0,0 +1,176 @@
+        });
+        expect(await ethers.provider.getBalance(instance.address)).to.eq(
+          initContractBalance.add(ethers.utils.parseEther('10.0')),
+        );
```
My thinking was to ensure everything worked correctly every step of the way, although I see now it is really unnecessary.

ItsNickBarry commented on test/utils/UintUtils.ts:
```diff
@@ -28,5 +28,80 @@ describe('UintUtils', function () {
+    describe('#toHexString', function () {
+      describe('(uint256)', function () {
+        it('returns 0 if input is 0', async () => {
+          const zeroNumber = ethers.constants.Zero;
```
No need for this variable - pass constant value directly.

ItsNickBarry commented:
> Should also be possible to replace the `while` loop in `UintUtils#toHexString` with this `for` loop:
> 
> ```
> for (uint256 temp = value; temp != 0; temp >>=8) {
>   length++;
> }
> ```
> 
> Check gas difference after implementing tests.

Gas report results:
| | Min | Max | Avg |
|-|-|-|-|
| *Before* | 23260 | 26520 | 24434 |
| *After* | 23249 | 26454 | 24403 |

Further optimization might be possible if a most-significant-bit function is used.

ItsNickBarry commented:
@NouDaimon `hardhat-abi-exporter` was recently updated to avoid deleting files that are not ABIs when clearing the output directory.  SolidState used to have a task override to copy and replace some files in the `abi` directory as a workaround for the previous behavior.  I removed that override when I updated `hardhat-abi-exporter`, but you still have the old version installed.  Please run `yarn install` and restore the deleted `abi/README.md` and `abi/package.json` files.

ItsNickBarry commented:
@NouDaimon I've restored the files myself, but make sure you run `yarn install`.  Also, I recommend adding files individually with `git add`, rather than using `git add .` or `git add -A` as most people seem to do.  To go even further, I recommend `git add -p` to choose only the specific changes you want within each file (this presents an opportunity to review all of your own changes before pushing).

ItsNickBarry commented on contracts/utils/UintUtils.sol:
```diff
@@ -63,7 +63,7 @@ library UintUtils {
         }
 
-        require(value == 0, 'Strings: hex length insufficient');
+        require(value == 0, 'UintUtils: hex length insufficient');
```
Need to update the corresponding test as well.

ItsNickBarry commented on test/utils/AddressUtils.ts:
```diff
@@ -137,18 +321,26 @@ describe('AddressUtils', async () => {
           ]);
 
-          await mock.mock.fn.returns(true);
+          //await mock.mock.fn.returns(true);
```
This allows the test to pass because the call reverts for an unrelated reason.  For now, we will use a function on `AddressUtilsMock` as you mention in your comment below.

ItsNickBarry replied:
See latest commit.

ItsNickBarry commented on contracts/utils/AddressUtilsMock.sol:
```diff
@@ -0,0 +1,60 @@
+        return target.functionCallWithValue(data, value, error);
+    }
+
+    function revertTest() external pure {
+        revert();
+    }
```
Remove this function, and instead simulate calls that revert like this:

```js
const mock = await deployMockContract(deployer, [
  'function fn () external payable returns (bool)',
]);

await mock.mock.fn.revertsWithReason(revertReason);
// or
await mock.mock.fn.reverts();
```

ItsNickBarry replied:
I standardized the tests across all similar functions, and removed this myself in the process.  No action needed.
